pulsar-io/nifi/src/main/java/org/apache/pulsar/io/nifi/NiFiSource.java

```diff
@@ -0,0 +1,185 @@
+            runnerThread.interrupt();
+            runnerThread.join();
+            runnerThread = null;
+        }
+    }
+
+    protected void start() throws Exception {
+        runnerThread = new Thread(new ReceiveRunnable());
+        runnerThread.setName("Nifi Source Thread");
+        runnerThread.start();
+    }
+
+
+    class ReceiveRunnable implements Runnable {
+
+        public ReceiveRunnable() {
+
+        }
+
+        @Override
+        public void run() {
+            try {
+                final SiteToSiteClient client = new SiteToSiteClient.Builder().fromConfig(clientConfig).build();
+                try {
+                    while (isRunning) {
+                        final Transaction transaction = client.createTransaction(TransferDirection.RECEIVE);
+                        DataPacket dataPacket = transaction.receive();
+                        if (dataPacket == null) {
+                            // no data available. Wait a bit and try again
+                            try {
+                                Thread.sleep(waitTimeMs);
+                            } catch (InterruptedException e) {
+                                log.warn("Failed to thread sleep milliseconds " + waitTimeMs, e);
+                            }
+                            continue;
+                        }
+
+                        final List<NiFiDataPacket> dataPackets = Lists.newArrayList();
+                        do {
+                            // Read the data into a byte array and wrap it along with the attributes
+                            // into a NiFiDataPacket.
+                            final InputStream inStream = dataPacket.getData();
+                            final byte[] data = new byte[(int) dataPacket.getSize()];
+                            StreamUtils.fillBuffer(inStream, data);
+
+                            final Map<String, String> attributes = dataPacket.getAttributes();
+                            final NiFiDataPacket NiFiDataPacket = new StandardNiFiDataPacket(data, attributes);
+                            dataPackets.add(NiFiDataPacket);
+                            dataPacket = transaction.receive();
+                        } while (dataPacket != null);
+
+                        // Confirm transaction to verify the data
+                        transaction.confirm();
+
+                        for (NiFiDataPacket dp : dataPackets) {
+                            consume(new NiFiRecord(dp, transaction));
+                        }
+
+                        transaction.complete();
+                    }
+                } finally {
+                    try {
+                        client.close();
+                    } catch (final IOException ioe) {
+                        log.warn("Failed to close client", ioe);
+                    }
+                }
+            } catch (final IOException ioe) {
+                log.warn("Failed to receive data from NiFi", ioe);
```

Comment:
does Nifi provide any `abort` method to abort a transaction? If it does, don't we need to abort the transaction when exception is thrown?

Reply:
> does Nifi provide any `abort` method to abort a transaction? If it does, don't we need to abort the transaction when exception is thrown?

Nifi provide transaction.error()